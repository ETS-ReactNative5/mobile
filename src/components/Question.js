import React, { Component } from 'react';
import {
  Dimensions,
  View
} from 'react-native';
import EStyleSheet from 'react-native-extended-stylesheet'
import StyledMarkdown from './StyledMarkdown'
import { connect } from 'react-redux'
import { setState } from '../actions/index'

//questionContainerHeight is stored in redux because the swipeable component needs
//to be absolutely positioned outside of the ContainerPanel for Android
//This is to help calculate that height

const mapStateToProps = (state, ownProps) => ({
  questionContainerHeight: state.classifier.questionContainerHeight[ownProps.workflowID] || 0,
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  setQuestionContainerHeight(height) {
    dispatch(setState(`classifier.questionContainerHeight.${ownProps.workflowID}`, height))
  },
})

export class Question extends Component {
  render() {
    return (
      <View style={[styles.questionContainer, { height: this.props.questionContainerHeight }]}>
        <View style={styles.question}>
          <StyledMarkdown
            extraCSS={ 'p {font-size: 14px; font-weight: 500; margin: 5px 0 0;}' }
            markdown={ this.props.question }
            width={ Dimensions.get('window').width - 100 }
            onResize={ (newHeight) => this.props.setQuestionContainerHeight(newHeight) }
          />
        </View>
      </View>
    )
  }
}

const styles = EStyleSheet.create({
  questionContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'stretch',
    marginTop: 10,
    marginHorizontal: 20,
  },
  question: {
    flex: 1,
    width: '100%',
  },
})

Question.propTypes = {
  question: React.PropTypes.string,
  workflowID: React.PropTypes.string,
  setQuestionContainerHeight: React.PropTypes.func,
  questionContainerHeight: React.PropTypes.number,
}

export default connect(mapStateToProps, mapDispatchToProps)(Question)
