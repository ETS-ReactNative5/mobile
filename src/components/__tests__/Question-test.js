import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
jest.mock('WebView', () => 'WebView')
import { Question } from '../Question'

it('renders correctly', () => {
  const tree = renderer.create(
    <Question
      question={'Is this this?'}
      questionContainerHeight={25}
      setQuestionContainerHeight={jest.fn}
      />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
