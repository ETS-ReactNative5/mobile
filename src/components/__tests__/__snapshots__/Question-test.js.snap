// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly (without task help) 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "height": 25,
      },
    ]
  }
>
  <View
    style={undefined}
  >
    <WebView
      automaticallyAdjustContentInsets={true}
      domStorageEnabled={true}
      javaScriptEnabled={true}
      onNavigationStateChange={[Function]}
      onShouldStartLoadWithRequest={[Function]}
      scalesPageToFit={true}
      source={
        Object {
          "baseUrl": "about:blank",
          "html": "<!DOCTYPE html>
<html>
    <head>
        <meta charset=\\"utf-8\\">
        <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
        <meta name=\\"viewport\\" content=\\"initial-scale=1, maximum-scale=1\\">
        <meta name=\\"format-detection\\" content=\\"telephone=no\\">
        <meta name=\\"format-detection\\" content=\\"date=no\\">
        <meta name=\\"format-detection\\" content=\\"address=no\\">
        <meta name=\\"format-detection\\" content=\\"email=no\\">

        <title></title>


        <style type=\\"text/css\\">
            body, html, #height-wrapper {
                margin: 0;
                padding: 0;
            }
            #height-wrapper {
                position: absolute;
                top: 0;
                left: 0;
                right: 0;
            }
            body {
                font-family: \\"Open Sans\\", \\"Gill Sans\\", Arial, sans-serif;
            }
            b, strong {
                font-family: \\"Open Sans\\", \\"Gill Sans\\", Arial, sans-serif;
                font-weight: bold;
            }
            h1, h2, h3, h4, h5, h6 {
                margin-top: 10px;
                font-family: \\"Open Sans\\", \\"Gill Sans\\", Arial, sans-serif;
                font-weight: 500;
            }
            img {
              height: auto;
              width: auto;
              max-width: 300px;
              max-height: 300px;
            }
            p {
              margin: 0;
            }
            p {font-size: 14px; font-weight: 500; margin: 5px 0 0;}
        </style>
    </head>
    <body>
        <p>Is this this?</p>


    <script>

    ;(function() {
    var wrapper = document.createElement(\\"div\\");
    wrapper.id = \\"height-wrapper\\";
    while (document.body.firstChild) {
        wrapper.appendChild(document.body.firstChild);
    }
    document.body.appendChild(wrapper);
    var i = 0;
    function updateHeight() {
        document.title = wrapper.clientHeight;
        window.location.hash = ++i;
    }
    updateHeight();
    window.addEventListener(\\"load\\", function() {
        updateHeight();
        setTimeout(updateHeight, 1000);
    });
    window.addEventListener(\\"resize\\", updateHeight);
    }());

    </script>
    </body>
</html>",
        }
      }
      style={
        Array [
          undefined,
          Object {
            "height": 0,
            "width": 650,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`renders correctly with task help 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "height": 25,
      },
    ]
  }
>
  <View
    style={undefined}
  >
    <WebView
      automaticallyAdjustContentInsets={true}
      domStorageEnabled={true}
      javaScriptEnabled={true}
      onNavigationStateChange={[Function]}
      onShouldStartLoadWithRequest={[Function]}
      scalesPageToFit={true}
      source={
        Object {
          "baseUrl": "about:blank",
          "html": "<!DOCTYPE html>
<html>
    <head>
        <meta charset=\\"utf-8\\">
        <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
        <meta name=\\"viewport\\" content=\\"initial-scale=1, maximum-scale=1\\">
        <meta name=\\"format-detection\\" content=\\"telephone=no\\">
        <meta name=\\"format-detection\\" content=\\"date=no\\">
        <meta name=\\"format-detection\\" content=\\"address=no\\">
        <meta name=\\"format-detection\\" content=\\"email=no\\">

        <title></title>


        <style type=\\"text/css\\">
            body, html, #height-wrapper {
                margin: 0;
                padding: 0;
            }
            #height-wrapper {
                position: absolute;
                top: 0;
                left: 0;
                right: 0;
            }
            body {
                font-family: \\"Open Sans\\", \\"Gill Sans\\", Arial, sans-serif;
            }
            b, strong {
                font-family: \\"Open Sans\\", \\"Gill Sans\\", Arial, sans-serif;
                font-weight: bold;
            }
            h1, h2, h3, h4, h5, h6 {
                margin-top: 10px;
                font-family: \\"Open Sans\\", \\"Gill Sans\\", Arial, sans-serif;
                font-weight: 500;
            }
            img {
              height: auto;
              width: auto;
              max-width: 300px;
              max-height: 300px;
            }
            p {
              margin: 0;
            }
            p {font-size: 14px; font-weight: 500; margin: 5px 0 0;}
        </style>
    </head>
    <body>
        <p>Is this this?</p>


    <script>

    ;(function() {
    var wrapper = document.createElement(\\"div\\");
    wrapper.id = \\"height-wrapper\\";
    while (document.body.firstChild) {
        wrapper.appendChild(document.body.firstChild);
    }
    document.body.appendChild(wrapper);
    var i = 0;
    function updateHeight() {
        document.title = wrapper.clientHeight;
        window.location.hash = ++i;
    }
    updateHeight();
    window.addEventListener(\\"load\\", function() {
        updateHeight();
        setTimeout(updateHeight, 1000);
    });
    window.addEventListener(\\"resize\\", updateHeight);
    }());

    </script>
    </body>
</html>",
        }
      }
      style={
        Array [
          undefined,
          Object {
            "height": 0,
            "width": 650,
          },
        ]
      }
    />
  </View>
  <View>
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      hitSlop={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onPress={[Function]}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          undefined,
          Object {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      ÔÅô
    </Text>
    <View>
      <Modal
        animationType="fade"
        hardwareAccelerated={false}
        onRequestClose={[Function]}
        transparent={true}
        visible={false}
      >
        <View
          style={undefined}
        >
          <RCTScrollView
            contentContainerStyle={undefined}
            style={undefined}
          >
            <View>
              <View
                style={undefined}
              >
                <WebView
                  automaticallyAdjustContentInsets={true}
                  domStorageEnabled={true}
                  javaScriptEnabled={true}
                  onNavigationStateChange={[Function]}
                  onShouldStartLoadWithRequest={[Function]}
                  scalesPageToFit={true}
                  source={
                    Object {
                      "baseUrl": "about:blank",
                      "html": "<!DOCTYPE html>
<html>
    <head>
        <meta charset=\\"utf-8\\">
        <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
        <meta name=\\"viewport\\" content=\\"initial-scale=1, maximum-scale=1\\">
        <meta name=\\"format-detection\\" content=\\"telephone=no\\">
        <meta name=\\"format-detection\\" content=\\"date=no\\">
        <meta name=\\"format-detection\\" content=\\"address=no\\">
        <meta name=\\"format-detection\\" content=\\"email=no\\">

        <title></title>


        <style type=\\"text/css\\">
            body, html, #height-wrapper {
                margin: 0;
                padding: 0;
            }
            #height-wrapper {
                position: absolute;
                top: 0;
                left: 0;
                right: 0;
            }
            body {
                font-family: \\"Open Sans\\", \\"Gill Sans\\", Arial, sans-serif;
            }
            b, strong {
                font-family: \\"Open Sans\\", \\"Gill Sans\\", Arial, sans-serif;
                font-weight: bold;
            }
            h1, h2, h3, h4, h5, h6 {
                margin-top: 10px;
                font-family: \\"Open Sans\\", \\"Gill Sans\\", Arial, sans-serif;
                font-weight: 500;
            }
            img {
              height: auto;
              width: auto;
              max-width: 300px;
              max-height: 300px;
            }
            p {
              margin: 0;
            }
            p{margin-top: 10px;}
        </style>
    </head>
    <body>
        <p>Halp!</p>


    <script>

    ;(function() {
    var wrapper = document.createElement(\\"div\\");
    wrapper.id = \\"height-wrapper\\";
    while (document.body.firstChild) {
        wrapper.appendChild(document.body.firstChild);
    }
    document.body.appendChild(wrapper);
    var i = 0;
    function updateHeight() {
        document.title = wrapper.clientHeight;
        window.location.hash = ++i;
    }
    updateHeight();
    window.addEventListener(\\"load\\", function() {
        updateHeight();
        setTimeout(updateHeight, 1000);
    });
    window.addEventListener(\\"resize\\", updateHeight);
    }());

    </script>
    </body>
</html>",
                    }
                  }
                  style={
                    Array [
                      undefined,
                      Object {
                        "height": 0,
                        "width": 670,
                      },
                    ]
                  }
                />
              </View>
              <View
                accessibilityComponentType={undefined}
                accessibilityLabel={undefined}
                accessibilityTraits={undefined}
                accessible={true}
                collapsable={undefined}
                hasTVPreferredFocus={undefined}
                hitSlop={undefined}
                isTVSelectable={true}
                nativeID={undefined}
                onLayout={undefined}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID={undefined}
                tvParallaxProperties={undefined}
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  style={undefined}
                >
                  Close
                </Text>
              </View>
            </View>
          </RCTScrollView>
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                undefined,
                Object {
                  "fontFamily": "FontAwesome",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            ÔÄç
          </Text>
        </View>
      </Modal>
    </View>
  </View>
</View>
`;
